cpu/sam0_common/periph/adc.c

```diff
@@ -120,33 +134,73 @@ static int _adc_configure(adc_res_t res)
     /* Power on */
 #ifdef CPU_SAML1X
     MCLK->APBCMASK.reg |= MCLK_APBCMASK_ADC;
+#endif
+#ifdef MCLK_APBDMASK_ADC0
+    if (ADC_DEV == ADC0) {
+        MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC0;
+    } else {
+        MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC1;
+    }
 #else
     MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC;
 #endif
+
+#ifdef ADC0_GCLK_ID
     /* GCLK Setup */
-    GCLK->PCHCTRL[ADC_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN_GCLK0;
-    /* Set Voltage Reference */
-    ADC->REFCTRL.reg = ADC_REF_DEFAULT;
+    if (ADC_DEV == ADC0) {
+        GCLK->PCHCTRL[ADC0_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+            GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
+    }
+    else {
+        GCLK->PCHCTRL[ADC1_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+            GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
+    }
+    /* Configure CTRLA & CTRLB Register */
+    ADC_DEV->CTRLA.reg = ADC_PRESCALER;
+    ADC_DEV->CTRLB.reg |= res;
+#else
+    /* GCLK Setup */
+    GCLK->PCHCTRL[ADC_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
     /* Configure CTRLB & CTRLC Register */
-    ADC->CTRLB.reg = ADC_PRESCALER;
-    ADC->CTRLC.reg |= res;
+    ADC_DEV->CTRLB.reg = ADC_PRESCALER;
+    ADC_DEV->CTRLC.reg |= res;
+#endif
```

Comment:
Can you move this block to a `_setup_clock()` helper function

Reply:
Done.